cdc: add elastic CPU control to CDC event processing

jayshrivastava opened the pull request:
### admission: make Pacer type available in SQL server config
Currently, the Pacer type is only used within KV, but will be used by SQL
in future changes. For example, code for encoding/decoding CDC events resides
in distSQL and is CPU intensive, so there is a plan to integrate admission
control to it in (https://github.com/cockroachdb/cockroach/issues/90089).
This change makes the Pacer type available to the SQL layer via the
`execinfra.ServerConfig`.

Because the Pacer was previously only used by KV, it lived in the `kvadmission`
package. Since this change makes it available outside of KV, it is moved to
the `admission` package.

Furthermore, this change adds a new method,
`ElasticCPUGrantCoordinator.NewPacer`, to instantiate new Pacer structs.
Since the `ElasticCPUGrantCoordinator` implements several features not relevant
to SQL, this change passes the coordinator to the SQL server config as
the interface `PacerMaker`, which makes only the `NewPacer` method accessible.

Currently tenant servers do not create grant coordinators for admission
control. This change retains that behavior, except it passes a `nil`
`ElasticCPUGrandCoordinator` which creates `nil`/noop Pacers. Adding these
coordinators to tenant servers is a change outside the scope of this commit and
is left as a `TODO`.

Release note: None

### cdc: add elastic CPU control to CDC event processing
Previously, the CPU-bound work of CDC event processing (encoding /
decoding rows) had the potential to consume a lot of CPU and
disrupt foreground SQL traffic. This changes adds elastic CPU control
to event processing so that it does not use excessive CPU and
starve foreground traffic.

This change also adds two new, non-public cluster settings, which control
enabling/disabling CPU control for CDC event processing and controlling
the requested grant size measured in CPU time.

Fixes: https://github.com/cockroachdb/cockroach/issues/90089

Release note: None

### roachtest: add initial scan only case to elastic cdc
Previously, this roachtest would not test changefeeds running with
`initial_scan_only`. This option tends to have a significant impact on
foreground latency due to high CPU usage, thus it should be included
in this test which measures CPU usage and foreground latency while
changefeeds are running.

Release note: None

jayshrivastava committed:
cdc: add elastic CPU control to CDC event processing

Previously, the CPU-bound work of CDC event processing (encoding /
decoding rows) had the potential to consume a lot of CPU and
disrupt foreground SQL traffic. This changes adds elastic CPU control
to event processing so that it does not use excessive CPU and
starve foreground traffic.

This change also adds a new, non-public cluster setting, which controls
enabling/disabling CPU control for CDC event processing and controlling
the requested grant size measured in CPU time.

Fixes: https://github.com/cockroachdb/cockroach/issues/90089

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/91554)
<!-- Reviewable:end -->

jayshrivastava commented:
### Before 
- value Encoding took 18.89% of CPU time
- CPU was near 100% utilized
- SQL service latency well **over** 100ms, peaking above 1s

<img width="1470" alt="image (2)" src="https://user-images.githubusercontent.com/18633281/200681622-a4ef38f1-0a0e-4541-86e7-f0ebb3529d0b.png">

<img width="1130" alt="image (4)" src="https://user-images.githubusercontent.com/18633281/200682357-eca3a536-b61a-4124-8613-1682043d0b48.png">





### After 
- value Encoding took 8.27% of CPU time
- CPU at ~70% utilization
- SQL service latency well **under** 100ms
<img width="1477" alt="image (3)" src="https://user-images.githubusercontent.com/18633281/200681697-594bfe0c-c9d3-45c9-9787-5c76a87dbcde.png">
<img width="1520" alt="image" src="https://user-images.githubusercontent.com/18633281/200682756-6d4ce85a-231b-470e-89b3-9c26fa3425d1.png">

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91554)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[`pkg/ccl/changefeedccl/changefeedbase/settings.go` line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGOG4cyEI88iDs51Xv1:-NGOG4cyEI88iDs51Xv2:b5v6ufw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a077f74fdba8fbb2b14bec07285157b29935d858/pkg/ccl/changefeedccl/changefeedbase/settings.go#L270)):*
> ```Go
> 		"before consuming events. after fully utilizing this CPU time, it will"+
> 		"request more",
> 	20*time.Millisecond,
> ```

I'm unsure what a reasonable default here would be. Event encoding / decoding time depends on the schema of the row. We also want don't want `Pace` to wait for quota on every loop iteration. 


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91554)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava and @miretskiy)

___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGRmyxD8qtOtl3-d_VS):*
nit: typo "This change makes the controller" 
___
*[`pkg/ccl/changefeedccl/event_processing.go` line 604 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGRow3yEkBZbNoMMYMr:-NGRow3yEkBZbNoMMYMs:bpsszcv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a077f74fdba8fbb2b14bec07285157b29935d858/pkg/ccl/changefeedccl/event_processing.go#L604)):*
> ```Go
> 				return c.setWorkerError(err)
> 			}
> 			c.decInFlight()
> ```

Minor: I'm assuming `pacer.Pace` and `consumer.ConsumeEvent` don't normally return an err, however, if they do, it would be better to call `c.decInFlight()` before returning. Otherwise, the count will never decrease. 
___
*[`pkg/ccl/changefeedccl/changefeedbase/settings.go` line 245 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGRoNl7B2lHEYPkE9-4:-NGRoNl7B2lHEYPkE9-5:b-1ymujn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a077f74fdba8fbb2b14bec07285157b29935d858/pkg/ccl/changefeedccl/changefeedbase/settings.go#L245)):*
> ```Go
> 	"changefeed.event_consumer_workers",
> 	"the number of workers to use when processing events: <0 disables, "+
> 		"0 assigns a reasonable default, >0 assigns the setting value. for expirimental/core "+
> ```

nit: expirimental -> experimental
___
*[`pkg/ccl/changefeedccl/changefeedbase/settings.go` line 254 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGRrd3M8b3Yf1-1Pr-P:-NGRrd3M8b3Yf1-1Pr-Q:b-h9huv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a077f74fdba8fbb2b14bec07285157b29935d858/pkg/ccl/changefeedccl/changefeedbase/settings.go#L254)):*
> ```Go
> 	settings.TenantWritable,
> 	"changefeed.event_consumer_worker_queue_size",
> 	"if changefeed.event_consumer_workers is enabled, this setting sets the maxmimum number of events"+
> ```

nit same issue here - add space between events and " -> of events "
___
*[`pkg/ccl/changefeedccl/changefeedbase/settings.go` line 267 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGRrHbaE9Ka2qxzFd1w:-NGRrHbaE9Ka2qxzFd1x:b-x9gvi4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a077f74fdba8fbb2b14bec07285157b29935d858/pkg/ccl/changefeedccl/changefeedbase/settings.go#L267)):*
> ```Go
> 	settings.TenantWritable,
> 	"changefeed.event_consumer_pacer_request_size",
> 	"an event consumer worker will perform a blocking request for CPU time"+
> ```

nit CPU time" -> CPU time " - add space after time see generated docs it becomes "timebefore". Also on the next line after will"
___
*[`pkg/ccl/changefeedccl/changefeedbase/settings.go` line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGOG4cyEI88iDs51Xv1:-NGRq6spCK1KUoSWcIXn:b-a5d04m) ([raw file](https://github.com/cockroachdb/cockroach/blob/a077f74fdba8fbb2b14bec07285157b29935d858/pkg/ccl/changefeedccl/changefeedbase/settings.go#L270)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

I'm unsure what a reasonable default here would be. Event encoding / decoding time depends on the schema of the row. We also want don't want `Pace` to wait for quota on every loop iteration. 
</blockquote></details>

minor: 9ms is likely a better default (assuming testing shows it works well). That is just under the length of the goroutine preemption time so on a busy system (which is when the Pacer will introduce pauses) it will cooperatively give up the CPU rather than waiting for preemption.


<!-- Sent from Reviewable.io -->

miretskiy commented:
> I'm unsure what a reasonable default here would be. Event encoding / decoding time depends on the schema of the row. We also want don't want `Pace` to wait for quota on every loop iteration.

I suggest that we run tpcc or even better, tpc-e.  elastic limiter already keeps track of few metrics (requests/returned); what would be nice is to add a histogram metric to keep track of how much real time was used.  Alternatively, add a metric to changefeed code to capture amount of time spent performing encoding + emit into sink (since we may do compression).  At any rate, whatever metric is used -- take 95th percentile as the configuration parameter.

miretskiy commented on pkg/ccl/changefeedccl/event_processing.go:
```diff
@@ -538,7 +591,16 @@ func (c *parallelEventConsumer) workerLoop(
+			// Request CPU time to use for event consumption, block if this time is
+			// unavailable. If there is unused CPU time left from the last call to
+			// Pace, then use that time instead of blocking.
+			if err := pacer.Pace(ctx); err != nil {
```
@irfansharif should review this usage -- if it's sane, etc (it does look okay to me).

irfansharif reviewed (changes requested):
The results look great! Few minor notes, I'll leave the rubber stamping to others. 
- Can we address the TODO here as part of this patch? Maybe every time `iters % 2 == 0` run the changefeed with `initial_scan = 'only'` instead.
https://github.com/cockroachdb/cockroach/blob/3d4f61a0bbb8a3ac08c6c5cafdb320238c284ad5/pkg/cmd/roachtest/tests/admission_control_elastic_cdc.go#L127-L131

- `kvadmission` is supposed to only house admission control utilities for the KV package. It's a thin layer between `pkg/kv/kvserver` and `pkg/admission`. The only reason the `Pacer` type was found there was because the only use of it was under `pkg/kv` but that's no longer the case. The only reason `kvadmission` is depended on in `pkg/server` is because it's an injected dependency. Lets improve it all by:
  - Moving `kvadmission.Pacer` into `pkg/util/admission`.
  - The way to construct a pacer should be `.NewPacer(unit time.Duration, wi admission.WorkInfo)` hanging off the `ElasticCPUGrantCoordinator` type. You should have access to this type here. 
https://github.com/cockroachdb/cockroach/blob/3ca64eb370ac53f5911914b35b90718489704558/pkg/server/server.go#L724
  - What should we actually hang off of `type sqlServerArgs struct`? kvadmission.Controller is too much, it's doing too many things that are irrelevant to SQL. We just want the ability to construct a pacer. Feel free to define a narrow `type admission.PacerConstructor interface` with just the `NewPacer` signature above that the `ElasticCPUGrantCoordinator` tacitly implements.
  - If we dropped the dependency on kvadmission.Controller, we wouldn't need special casing for the missing tenant integration. Here for example we have a link to some admission queue that also works in tenant code. We should be able to do similar plumbing for this elastic CPU controller as well:
https://github.com/cockroachdb/cockroach/blob/3ca64eb370ac53f5911914b35b90718489704558/pkg/server/server.go#L848
- Add a cluster setting to disable this specific form of pacing, the narrow equivalent of `kvadmission.rangefeed_catchup_scan_elastic_control.enabled`. Default it to true, but during incidents it's always good to have a kill-switch.
- Do we need the `wrappedPacer` type at all? A nil `Pacer` is something you can call .Close and .Pace on without it doing anything -- try it! We use this property when the cluster setting above is disabled.
- In the pacer error path, I would recommend a different fallback: just log semi-infrequently and continue processing. Failing to provide latency isolation is less bad some times than failing parallel workers altogether. See what we're doing here:
https://github.com/cockroachdb/cockroach/blob/a042b60ae1da66b195f4d2537e3a147473a6778a/pkg/kv/kvserver/rangefeed/catchup_scan.go#L168

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-:-NGSzS5GCSMsMimGThif:bglwogy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @irfansharif, @jayshrivastava, and @miretskiy)

___
*[`pkg/ccl/changefeedccl/event_processing.go` line 568 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGO9SsDsSqyysn9oGN0-r2-568:-NGTB3w_7Zl3_KkX6zZf:b598cqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ca64eb370ac53f5911914b35b90718489704558/pkg/ccl/changefeedccl/event_processing.go#L568)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I wonder if controller itself could have a "NewPacer" method?  It can even return a no-op pacer.  
</blockquote></details>

+1 to NewPacer, perhaps as an interface, and not hanging off the kvadmission.Controller thing (see top-level comment). A nil pacer is still something you can call .Close or .Pace on BTW, see top-level comment. That's a "no-op" pacer.
___
*[`pkg/ccl/changefeedccl/event_processing.go` line 571 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGO9SsDsSqyysn9oGN0-r2-571:-NGTBHxe6O9K83CXEj9G:bxl5slj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ca64eb370ac53f5911914b35b90718489704558/pkg/ccl/changefeedccl/event_processing.go#L571)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

is this a test only thing(that ac is nil)?
</blockquote></details>

We shouldn't need this, see top-level comment.
___
*[`pkg/ccl/changefeedccl/event_processing.go` line 597 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGO9SsDsSqyysn9oGN0-r2-597:-NGTAyhO88FG1rAgFTui:bqma9du) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ca64eb370ac53f5911914b35b90718489704558/pkg/ccl/changefeedccl/event_processing.go#L597)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

@irfansharif should review this usage -- if it's sane, etc (it does look okay to me).
</blockquote></details>

See comment above around error handling. I'll defer to you two. Could we call `if err := pacer.Pace(ctx); err != nil {` outside the switch block? Just in case we add new case statements here.
___
*[`pkg/ccl/changefeedccl/changefeedbase/settings.go` line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGOG4cyEI88iDs51Xv1:-NGTBYTX5bpJtCKAnmZs:b-mu6ttm) ([raw file](https://github.com/cockroachdb/cockroach/blob/a077f74fdba8fbb2b14bec07285157b29935d858/pkg/ccl/changefeedccl/changefeedbase/settings.go#L270)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

minor: 9ms is likely a better default (assuming testing shows it works well). That is just under the length of the goroutine preemption time so on a busy system (which is when the Pacer will introduce pauses) it will cooperatively give up the CPU rather than waiting for preemption.
</blockquote></details>

I actually think 9ms is too little for this tight loop -- it's better to do as much work as possible without interacting with AC too often. I'm not sure I understand what the practical difference is between cooperatively giving up the CPU rather than waiting for pre-emption, we'll still get prempted. BTW Andrew, pre-emption in the Go scheduler could also just immediately put it on another inactive core, or not even pre-empt it if there's no waiting work on that core. There are some scheduler traces in #unified-overload-control to back me up.

I'll recommend something similar to what @miretskiy is saying: run an experiment with TPC-C, or TPC-E and see what the aggregate CPU-time from this tight loop typically is (I would just log.Infof it -- quicker than a histogram). I suspect it's 100ms+. Pick something like 100ms or 50ms and call it a day. We can always re-evaluate defaults if need be. I also don't think this upfront estimate matters much in practice because we correct for this error term post-hoc.


<!-- Sent from Reviewable.io -->

jayshrivastava committed:
admission: make Pacer type available in SQL server config

Currently, the Pacer type is only used within KV, but will be used by SQL
in future changes. For example, code for encoding/decoding CDC events resides
in distSQL and is CPU intensive, so there is a plan to integrate admission
control to it in (https://github.com/cockroachdb/cockroach/issues/90089).
This change makes the Pacer type available to the SQL layer via the
`execinfra.ServerConfig`.

Because the Pacer was previously only used by KV, it lived in the `kvadmission`
package. Since this change makes it available outside of KV, it is moved to
the `admission` package.

Furthermore, this change adds a new method,
`ElasticCPUGrantCoordinator.NewPacer`, to instantiate new Pacer structs.
Since the `ElasticCPUGrantCoordinator` implements several features not relevant
to SQL, this change passes the coordinator to the SQL server config as
the interface `PacerMaker`, which makes only the `NewPacer` method accessible.

Currently tenant servers do not create grant coordinators for admission
control. This change retains that behavior, except it passes a `nil
ElasticCPUGrandCoordinator` which creates `nil`/noop Pacers. Adding these
coordinators to tenant servers is a change outside the scope of this commit and
is left as a `TODO`.

Release note: None

sumeerbhola reviewed (commented):
> Reviewed 10 of 17 files at r3, 1 of 3 files at r4, 3 of 7 files at r5, 5 of 5 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91554)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @jayshrivastava, and @miretskiy)

___
*[`pkg/ccl/changefeedccl/changefeedbase/settings.go` line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGOG4cyEI88iDs51Xv1:-NGY_OPi8Fj5jBIBKwSy:buzf660) ([raw file](https://github.com/cockroachdb/cockroach/blob/a077f74fdba8fbb2b14bec07285157b29935d858/pkg/ccl/changefeedccl/changefeedbase/settings.go#L270)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

I actually think 9ms is too little for this tight loop -- it's better to do as much work as possible without interacting with AC too often. I'm not sure I understand what the practical difference is between cooperatively giving up the CPU rather than waiting for pre-emption, we'll still get prempted. BTW Andrew, pre-emption in the Go scheduler could also just immediately put it on another inactive core, or not even pre-empt it if there's no waiting work on that core. There are some scheduler traces in #unified-overload-control to back me up.

I'll recommend something similar to what @miretskiy is saying: run an experiment with TPC-C, or TPC-E and see what the aggregate CPU-time from this tight loop typically is (I would just log.Infof it -- quicker than a histogram). I suspect it's 100ms+. Pick something like 100ms or 50ms and call it a day. We can always re-evaluate defaults if need be. I also don't think this upfront estimate matters much in practice because we correct for this error term post-hoc.
</blockquote></details>

+! to making this 50+ms, and deciding based on measurement, since interacting with AC is not super cheap.


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
Addressing Irfan's Comments: 
- Addressing the `TODO`
  - I think it's better to do this in a separate patch since there's also already some in flight work for `initial_scan_only` in a different, but similar, roachtest: https://github.com/cockroachdb/cockroach/pull/91394

- `kvadmission` / `admission` package improvements
  - I moved the `Pacer` to it's own file in the `admission package`
  - I added `ElasticCPUGrantCoordinator.NewPacer`
  - I added the new `PacerMaker` interface to hide some of the `ElasticCPUGrantCoordinator`s functionality
  - Unfortunately, the example you gave with the `sqlSQLResponseAdmissionQ` seems to be a noop with tenant servers. It only gets initialized for non-tenant servers in `pkg/server/server.go` and remains uninitialized in `pkg/server/tenant.go`. Every time it's used (such as below), it's wrapped with a `nil` check. I think it's best to keep this functionality until we initialize grant coordinators for tenant servers, which can be done in a separate patch. I left a TODO for this.https://github.com/cockroachdb/cockroach/blob/3ca64eb370ac53f5911914b35b90718489704558/pkg/sql/colflow/colrpc/inbox.go#L417-L423 https://github.com/cockroachdb/cockroach/blob/3ca64eb370ac53f5911914b35b90718489704558/pkg/sql/flowinfra/inbound.go#L214-L218

- Kill switch
  - I updated the existing setting to disable pacing if the value is set to 0

- Wrapped Pacer
  - I removed this type. You're right - the `nil` pacer can be used as a `noop` pacer. The scheme I implemented is that tenant servers will be bootstrapped with a nil `ElasticCPUGrantCoordinator`, whose `NewPacer` method returns a `nil` pacer. Then this pacer is used in CDC event processing. CDC event processing is not aware if the pacer is nil or not. Whenever we decide to initialize a coordinator for tenant servers, the CDC code won't require changes.

- Pacer Errors
  - I updated the code to log Pacer errors infrequently instead of returning them

@miretskiy, do we still want metrics for time spent encoding? You can see my comment below regarding 50ms. I did a small experiment with logging.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-:-NGc7RJl98GySVLxIxEO:b-shfxcb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @miretskiy, and @sumeerbhola)

___
*[`pkg/ccl/changefeedccl/event_processing.go` line 604 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGRow3yEkBZbNoMMYMr:-NGbh-bk4l0D3og4KL6J:b-hm6su4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a077f74fdba8fbb2b14bec07285157b29935d858/pkg/ccl/changefeedccl/event_processing.go#L604)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Minor: I'm assuming `pacer.Pace` and `consumer.ConsumeEvent` don't normally return an err, however, if they do, it would be better to call `c.decInFlight()` before returning. Otherwise, the count will never decrease. 
</blockquote></details>

I tried doing this and some tests deadlocked. Calling`c.decInFlight()` less often is definitely the more cautious approach. If there's a terminal error, the number of in flight events doesn't matter much. 
___
*[`pkg/ccl/changefeedccl/event_processing.go` line 119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGO9SsDsSqyysn9oGN0-r2-119:-NGc6VyH-7MQGSP1I76L:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ca64eb370ac53f5911914b35b90718489704558/pkg/ccl/changefeedccl/event_processing.go#L119)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

make this todo more specific -- it's about parquet.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/event_processing.go` line 124 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGO9SsDsSqyysn9oGN0-r2-124:-NGbeyf86V8j4EQwgeIf:baheaht) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ca64eb370ac53f5911914b35b90718489704558/pkg/ccl/changefeedccl/event_processing.go#L124)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Do you remember why this was disabled in the first place?
</blockquote></details>

This I remember doing a small fix for this, but we dropped it because the nprocs PR was big enough already and core changefeeds are still in beta. The problem is `changeAggregator.Next()` calls `ConsumeEvent` and immediately expects some data to be synchronously placed in the output buffer afterwards. Thus, async encoding/emitting doesn't work.
___
*[`pkg/ccl/changefeedccl/event_processing.go` line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGO9SsDsSqyysn9oGN0-r2-125:-NGbfx-qAlKv9D573yvB:bb77ufg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ca64eb370ac53f5911914b35b90718489704558/pkg/ccl/changefeedccl/event_processing.go#L125)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I believe you've updated settings description already?
</blockquote></details>

Yes.
___
*[`pkg/ccl/changefeedccl/event_processing.go` line 568 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGO9SsDsSqyysn9oGN0-r2-568:-NGc5vyj4UvaWZlELRHT:b6n2ohr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ca64eb370ac53f5911914b35b90718489704558/pkg/ccl/changefeedccl/event_processing.go#L568)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

+1 to NewPacer, perhaps as an interface, and not hanging off the kvadmission.Controller thing (see top-level comment). A nil pacer is still something you can call .Close or .Pace on BTW, see top-level comment. That's a "no-op" pacer.
</blockquote></details>

I refactored this quite a bit. We can have nil `PacerMaker` and nil `Pacer`, both of which basically operate as noops for tenant servers. More info in my top-level reply. 
___
*[`pkg/ccl/changefeedccl/event_processing.go` line 573 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGO9SsDsSqyysn9oGN0-r2-573:-NGbgTNU5F1b8NqhhiEN:b-1v7hk8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ca64eb370ac53f5911914b35b90718489704558/pkg/ccl/changefeedccl/event_processing.go#L573)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

this looks strange -- aren't you going to immediately close pacer?
</blockquote></details>

Yes. That was a mistake. It's fixed now.
___
*[`pkg/ccl/changefeedccl/event_processing.go` line 597 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGO9SsDsSqyysn9oGN0-r2-597:-NGc6Lc1DBxzCbEmfpLn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ca64eb370ac53f5911914b35b90718489704558/pkg/ccl/changefeedccl/event_processing.go#L597)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

See comment above around error handling. I'll defer to you two. Could we call `if err := pacer.Pace(ctx); err != nil {` outside the switch block? Just in case we add new case statements here.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/changefeedbase/settings.go` line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGOG4cyEI88iDs51Xv1:-NGbhxOD8PpGQc-nUO9x:b-5sup5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/a077f74fdba8fbb2b14bec07285157b29935d858/pkg/ccl/changefeedccl/changefeedbase/settings.go#L270)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

+! to making this 50+ms, and deciding based on measurement, since interacting with AC is not super cheap.
</blockquote></details>

So I did some testing as @irfansharif mentioned. I ran TPCC w/ 1000 warehouses with an initial scan on `tpcc.order_line`. Also tried a changefeed with the `UPDATED, DIFF, RESOLVED` options. The average CPU time per call to `ConsumeEvent` (measured with the `grunning` pkg) was 17391.80ns and 21452.73ns respectively. I think this measurement (and even a p95 measurement) depends a lot on the schema of the table and size of each row, so we should take it with a grain of salt. 

I was a bit confused about what you meant by aggregate CPU time. The cumulative CPU time would just be a function of how many events we need to process, which does not seem like a useful measurement. I measured this to be 219990339846ns and 274068743360ns respectively... Large numbers because we ran a changefeed on a large table.

I'm also confused about why we ran the experiment with TPCC. I feel that the CPU time spent encoding events would be the same regardless of an external workload running. Unless we include context switching overhead in the requested grant size for some reason... 

Nevertheless, it sounds like everyone's okay with 50ms. I updated the setting to default to 50ms.  It means we can process ~2500 events before requesting another grant from AC (well, events coming from `order_line` table since each row takes ~20000ns). At the end of the day, it's a default setting and it can be changed for a particular table/schema.
 
I also sanity checked that the graphs look similar with 50ms.
![image.png](https://files.reviewableusercontent.io/images/reviewable/18633281/aecHWERPZ-UFKbQa2mu_btjB/image.png)
___
*[`pkg/ccl/changefeedccl/changefeedbase/settings.go` line 266 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGO9SsDsSqyysn9oGN--r2-266:-NGbjkEr7wNqFO5i0sX8:bwv796e) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ca64eb370ac53f5911914b35b90718489704558/pkg/ccl/changefeedccl/changefeedbase/settings.go#L266)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

nit: wdyt about `changefeed.cpu.per_event_allocation` or something similar?
</blockquote></details>

I can do `changefeed.cpu.per_event_consumer_worker_allocation`. I don't want to use `per_event` since we will encode many events during that time. 
___
*[`pkg/ccl/changefeedccl/changefeedbase/settings.go` line 272 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGO9SsDsSqyysn9oGN--r2-272:-NGc7KEsAYsL6wk9vp6U:b-i0prpi) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ca64eb370ac53f5911914b35b90718489704558/pkg/ccl/changefeedccl/changefeedbase/settings.go#L272)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm okay keeping this setting; I'm not sure we need it withPublic -- it's very low level.
</blockquote></details>

Sure. I removed `withPublic`.


<!-- Sent from Reviewable.io -->

jayshrivastava commented on pkg/ccl/changefeedccl/event_processing.go:
I replied in my latest comment

miretskiy replied:
Having hard time finding it, I'm afraid.

jayshrivastava replied:
> I remember doing a small fix for this, but we dropped it because the nprocs PR was big enough already and core changefeeds are still in beta. The problem is `changeAggregator.Next()` calls ConsumeEvent and immediately expects some data to be synchronously placed in the output buffer afterwards. Thus, async encoding/emitting doesn't work.

miretskiy commented on pkg/ccl/changefeedccl/event_processing.go:
```diff
@@ -174,6 +218,8 @@ func newKVEventToRowConsumer(
 	knobs TestingKnobs,
 	topicNamer *TopicNamer,
+	pacer *admission.Pacer,
+	pacerLogEvery *log.EveryN,
```
you could use a single (global) everyN object for all pacers -- I doubt you need it per consumer.

jayshrivastava replied:
Done.

miretskiy commented on pkg/ccl/changefeedccl/event_processing.go:
```diff
@@ -100,22 +121,45 @@ func newEventConsumer(
+			}
+			pacer = cfg.PacerMaker.NewPacer(
+				pacerRequestUnit,
+				admission.WorkInfo{
```
This workInfo struct should probably have some sort of a tag (an op-string or whatnot) identifying the queue. I know that at this point, pacer does not export per queue stats, but it probably should.
A todo would be fine (in the admission code)

jayshrivastava replied:
I'm worried about what would happen if you tag a `WorkInfo` with one queue's info, but send it to another. Also, maybe queues should collect stats instead of the `Pacer` type, because they are underlying mechanism of the `Pacer type. 

It may make more sense to tag `WorkInfo` with the type/origin of the work and have queues collect metrics about the work coming in. We can group queues by their [granter](https://github.com/cockroachdb/cockroach/blob/c8c85358743be9c1de1c8765352f7269f1d98ff1/pkg/util/admission/work_queue.go#L226-L227). For example, collect stats about cdc work being granted CPU time. @sumeerbhola, would you mind sharing your opinion about this?

miretskiy commented on pkg/util/admission/grant_coordinator.go:
```diff
@@ -1014,3 +1014,25 @@ func (e *ElasticCPUGrantCoordinator) close() {
+// PacerMaker interface should be used in the SQL layer in lieu of
+// ElasticCPUGrantCoordinator because the ElasticCPUGrantCoordinator
+// implements several features which are irrelevant to SQL.
+type PacerMaker interface {
```
This sounds so medical!

jayshrivastava replied:
I was resisting the urge to call it `PaceMaker`.

miretskiy reviewed (commented):
cdc PR needs to be updated to no longer mention server/sql as the type of the change.

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91554)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @jayshrivastava, and @miretskiy)

___
*[`pkg/ccl/changefeedccl/changefeedbase/settings.go` line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGOG4cyEI88iDs51Xv1:-NGqiLL_8eNTgwt89ZqD:b-9m4f2h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a077f74fdba8fbb2b14bec07285157b29935d858/pkg/ccl/changefeedccl/changefeedbase/settings.go#L270)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

So I did some testing as @irfansharif mentioned. I ran TPCC w/ 1000 warehouses with an initial scan on `tpcc.order_line`. Also tried a changefeed with the `UPDATED, DIFF, RESOLVED` options. The average CPU time per call to `ConsumeEvent` (measured with the `grunning` pkg) was 17391.80ns and 21452.73ns respectively. I think this measurement (and even a p95 measurement) depends a lot on the schema of the table and size of each row, so we should take it with a grain of salt. 

I was a bit confused about what you meant by aggregate CPU time. The cumulative CPU time would just be a function of how many events we need to process, which does not seem like a useful measurement. I measured this to be 219990339846ns and 274068743360ns respectively... Large numbers because we ran a changefeed on a large table.

I'm also confused about why we ran the experiment with TPCC. I feel that the CPU time spent encoding events would be the same regardless of an external workload running. Unless we include context switching overhead in the requested grant size for some reason... 

Nevertheless, it sounds like everyone's okay with 50ms. I updated the setting to default to 50ms.  It means we can process ~2500 events before requesting another grant from AC (well, events coming from `order_line` table since each row takes ~20000ns). At the end of the day, it's a default setting and it can be changed for a particular table/schema.
 
I also sanity checked that the graphs look similar with 50ms.
![image.png](https://files.reviewableusercontent.io/images/reviewable/18633281/aecHWERPZ-UFKbQa2mu_btjB/image.png)

</blockquote></details>

Regarding "aggregate CPU-time", I'll give my take on it and @irfansharif may possibly differ.

Normally, the high CPU work that integrates with admission control takes the form of a request that is evaluated by doing a tight CPU intensive loop e.g. KV `BatchRequest` processing where the request is an ExportRequest. We can measure the goroutine running time of these, and have found that they frequently consume > 200ms for a single request. The admission.ElasticCPUWorkHandle was used to make these requests stop doing more after 100ms and return (with a resumption point that the caller would use to send the next request). The `CatchupIterator.CatchUpScan` is different in that it is not a request-response pattern, but one can measure the CPU time spent in that method since it loops and calls `outputFn`. The `Pacer` (that wraps the elastic handle and queue) is used to seek re-admission after every 100ms of CPU work.
So my notion of "aggregate CPU-time" for the initial scan case in this PR was similar to those: how much total CPU time do we spend in the goroutine that is consuming these events produced via the concurrent scans and calling `kvEventToRowConsumer.ConsumeEvent`. The effectiveness of using a 50ms granularity for AC reduces if the CPU time spent in this goroutine is typically less than 50ms.

An underlying assumption is that because of the concurrent scans this goroutine is typically not blocked in waiting state for an event -- if it was frequently blocked because the scans were not feeding it enough work, its latency impact on others would be inherently lower (in that it would not be competing as aggressively for CPU in the goroutine scheduler).


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91554)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @miretskiy, and @sumeerbhola)

___
*[`pkg/ccl/changefeedccl/changefeedbase/settings.go` line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGOG4cyEI88iDs51Xv1:-NGr0jU7856RSet9TwAm:b-bwmmvd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a077f74fdba8fbb2b14bec07285157b29935d858/pkg/ccl/changefeedccl/changefeedbase/settings.go#L270)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Regarding "aggregate CPU-time", I'll give my take on it and @irfansharif may possibly differ.

Normally, the high CPU work that integrates with admission control takes the form of a request that is evaluated by doing a tight CPU intensive loop e.g. KV `BatchRequest` processing where the request is an ExportRequest. We can measure the goroutine running time of these, and have found that they frequently consume > 200ms for a single request. The admission.ElasticCPUWorkHandle was used to make these requests stop doing more after 100ms and return (with a resumption point that the caller would use to send the next request). The `CatchupIterator.CatchUpScan` is different in that it is not a request-response pattern, but one can measure the CPU time spent in that method since it loops and calls `outputFn`. The `Pacer` (that wraps the elastic handle and queue) is used to seek re-admission after every 100ms of CPU work.
So my notion of "aggregate CPU-time" for the initial scan case in this PR was similar to those: how much total CPU time do we spend in the goroutine that is consuming these events produced via the concurrent scans and calling `kvEventToRowConsumer.ConsumeEvent`. The effectiveness of using a 50ms granularity for AC reduces if the CPU time spent in this goroutine is typically less than 50ms.

An underlying assumption is that because of the concurrent scans this goroutine is typically not blocked in waiting state for an event -- if it was frequently blocked because the scans were not feeding it enough work, its latency impact on others would be inherently lower (in that it would not be competing as aggressively for CPU in the goroutine scheduler).
</blockquote></details>

Thanks for the context! Fortunately, in this scenario, we have goroutines which run for much longer than 50ms, and are typically not blocked. 


<!-- Sent from Reviewable.io -->

miretskiy commented:
> Thanks for the context! Fortunately, in this scenario, we have goroutines which run for much longer than 50ms, and are typically not blocked.

True that they run for longer than 50ms -- they run forever, hopefully, -- but they certainly can get blocked (i.e. trying to emit to the sink -- which is blocked). I don't think this changes your answer though, does it?

miretskiy reviewed (approved):
Giving 👍  since I'm going to be leaving soon on vaca.
You still need to get +1 from Irfan/Sumeer.

jayshrivastava commented:
@miretskiy TYFR! I'll wait on another ✅ before merging. Also, you're right, the consumers get blocked on emitting. I meant to say that they don't get blocked waiting for events to come through from kv scans, as Sumeer mentioned.

sumeerbhola commented:
The initial scan also finishes at some point, yes? My mental model is clearly very fuzzy.

miretskiy commented:
Initial scan does finish -- but this same code processes rangefeed events
-- which never finish.

On Thu, Nov 17, 2022 at 1:50 PM sumeerbhola ***@***.***>
wrote:

> The initial scan also finishes at some point, yes? My mental model is
> clearly very fuzzy.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/91554#issuecomment-1319061031>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANA4FVHLSNAIQFKJMHUZKMLWIZ45RANCNFSM6AAAAAAR2YT2EE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

jayshrivastava committed:
roachtest: add initial scan only case to elastic cdc

Previously, this roachtest would not test changefeeds running with
`initial_scan_only`. This option tends to have a significant impact on
foreground latency due to high CPU usage, thus it should be included
in this test which measures CPU usage and foreground latency while
changefeeds are running.

Release note: None

jayshrivastava commented:
I pushed another commit which adds the `initial_scan_only` case, to address @irfansharif's TODO. The results are good:

Before (latest master as of commenting):
<img width="1536" alt="image (6)" src="https://user-images.githubusercontent.com/18633281/202568803-40fea457-4097-419e-9ec3-4fe06498d191.png">

After (my changes):
<img width="1531" alt="image (5)" src="https://user-images.githubusercontent.com/18633281/202568822-29b5ca69-fcf2-4125-bb2f-adcd6b722faf.png">

irfansharif reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>, only left nits below. Thanks a ton for working through it.

> I updated the existing setting to disable pacing if the value is set to 0

[mega nit] Mind just adding another setting instead? Something like `kvadmission.rangefeed_catchup_scan_elastic_control.enabled` as suggested above. It's easier to grep for `.enabled` when looking for cluster settings.

> Reviewed 8 of 17 files at r3, 1 of 3 files at r4, 1 of 7 files at r5, 9 of 9 files at r7, 3 of 6 files at r8, 1 of 3 files at r9, 2 of 2 files at r10, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-:-NHabT0lCuWFlq6mALBB:b-sfch87)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jayshrivastava, @miretskiy, and @sumeerbhola)

___
*[`pkg/ccl/changefeedccl/event_processing.go` line 134 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGO9SsDsSqyysn9oGN0-r9-134:-NHaTpXQ18gT6MhPgB4c:bhfoetv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d625d495bc539ca23ac7c46303cd4f420192501f/pkg/ccl/changefeedccl/event_processing.go#L134)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

I'm worried about what would happen if you tag a `WorkInfo` with one queue's info, but send it to another. Also, maybe queues should collect stats instead of the `Pacer` type, because they are underlying mechanism of the `Pacer type. 

It may make more sense to tag `WorkInfo` with the type/origin of the work and have queues collect metrics about the work coming in. We can group queues by their [granter](https://github.com/cockroachdb/cockroach/blob/c8c85358743be9c1de1c8765352f7269f1d98ff1/pkg/util/admission/work_queue.go#L226-L227). For example, collect stats about cdc work being granted CPU time. @sumeerbhola, would you mind sharing your opinion about this?
</blockquote></details>

I doubt we'll export per "user-of-Pacer" type statistics, so this change as is is probably just fine for now/immediate future. It's pretty clear from metrics today which process is using these elastic CPU tokens since there are only a handful of uses (backups, changefeed parallel workers, catch up scans, additional pebble compression threads). Can revisit if it becomes useful to segment.
___
*[`pkg/ccl/changefeedccl/event_processing.go` line 127 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NHaa2x1C3QKblO--Slv:-NHaa2x1C3QKblO--Slw:b-7ynjbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d3bd357769120ef23b8a3112bffc35141eb79ca2/pkg/ccl/changefeedccl/event_processing.go#L127)):*
> ```Go
> 		// Passing a nil Pacer is effectively a noop Pacer if
> 		// CPU control is disabled.
> 		var pacer *admission.Pacer = nil
> ```

[nit] The `= nil` is unnecessary.
___
*[`pkg/ccl/changefeedccl/changefeedbase/settings.go` line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NGOG4cyEI88iDs51Xv1:-NHaZ_1z0zpHEyStpb6X:bobks4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/a077f74fdba8fbb2b14bec07285157b29935d858/pkg/ccl/changefeedccl/changefeedbase/settings.go#L270)):*
<details><summary><i>Previously, jayshrivastava (Jayant Shrivastava) wrote…</i></summary><blockquote>

Thanks for the context! Fortunately, in this scenario, we have goroutines which run for much longer than 50ms, and are typically not blocked. 
</blockquote></details>

This all sounds right to me. And also, great results!
___
*[`pkg/cmd/roachtest/tests/admission_control_elastic_cdc.go` line 127 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NHaazUOE2kEcfEFzQyo:-NHaazUOE2kEcfEFzQyp:brgt7u1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e23bc3126c0a10d64976700c9977cc28d4a7be36/pkg/cmd/roachtest/tests/admission_control_elastic_cdc.go#L127)):*
> ```Go
> 								} else {
> 									createChangefeedStmt = "CREATE CHANGEFEED FOR tpcc.order_line, tpcc.stock, tpcc.customer " +
> 										"INTO 'null://' WITH initial_scan_only"
> ```

[nit] `initial_scan = 'only'` is the preferred form I think: https://github.com/cockroachdb/cockroach/pull/79324.
___
*[`pkg/server/tenant.go` line 880 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NHaZfUj8Pgc95we6EKl:-NHaZfUj8Pgc95we6EKm:bc3en6m) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc44f51e8993b3b49f8c218b4a43874a13a4c40f/pkg/server/tenant.go#L880)):*
> ```Go
> 	}
> 
> 	// TODO(jayant): generate admission.NewGrantCoordinators and pass them
> ```

Can you name me in this TODO instead? And it should reference NewGrantCoordinatorSQL instead -- which like you've seen, is not hooked up. Also, [nit] the `= nil` is unnecessary
___
*[`pkg/sql/execinfra/server_config.go` line 196 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NHaZLAl0B2jxg167heL:-NHaZLAl0B2jxg167heM:b-b411ym) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc44f51e8993b3b49f8c218b4a43874a13a4c40f/pkg/sql/execinfra/server_config.go#L196)):*
> ```Go
> 	RangeStatsFetcher eval.RangeStatsFetcher
> 
> 	PacerMaker admission.PacerMaker
> ```

Add a comment.
___
*[`pkg/util/admission/grant_coordinator.go` line 1020 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NHaVBwbBce7pLu-IRGK:-NHaVBwbBce7pLu-IRGL:bc5m9lg) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc44f51e8993b3b49f8c218b4a43874a13a4c40f/pkg/util/admission/grant_coordinator.go#L1020)):*
> ```Go
> // PacerMaker is used to construct a new admission.Pacer.
> //
> // PacerMaker interface should be used in the SQL layer in lieu of
> ```

Drop this paragraph, there's no need to talk about call sites which are ever changing. It's also talking a lot about the implementation which is distracting. Move this interface definition to admission.go which holds the other interfaces in this package. Call this PacerFactory instead (we have various "factory" interfaces elsewhere in CRDB) -- rename instances of "pacerMaker" to "admissionPacerFactory" when you do.
___
*[`pkg/util/admission/grant_coordinator.go` line 1024 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NHaVxqP2YmhqEwDGMSz:-NHaVxqP2YmhqEwDGMT-:bx040lf) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc44f51e8993b3b49f8c218b4a43874a13a4c40f/pkg/util/admission/grant_coordinator.go#L1024)):*
> ```Go
> // implements several features which are irrelevant to SQL.
> type PacerMaker interface {
> 	// NewPacer constructs a new admission.Pacer, which may be nil.
> ```

Drop this comment, it's a single method interface and the top-level comment is enough. Also, there's no need to scare users off by mentioning "nil". All methods support the nil receiver so it should be irrelevant to users.
___
*[`pkg/util/admission/grant_coordinator.go` line 1028 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NHaWK0U1vQL7shXTB7v:-NHaWK0U1vQL7shXTB7w:b-ve9joc) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc44f51e8993b3b49f8c218b4a43874a13a4c40f/pkg/util/admission/grant_coordinator.go#L1028)):*
> ```Go
> }
> 
> // NewPacer implements the PacerMaker interface.
> ```

Add one of these interface assertions somewhere

```
var _ PacerFactory = &ElasticCPUGrantCoordinator{}
```


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
Thanks for reviewing! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-:-NI3fcvABeogqBCSk8qo:b-irrk4g)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andrewbaptist, @irfansharif, @miretskiy, and @sumeerbhola)

___
*[`pkg/server/tenant.go` line 880 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/91554#-NHaZfUj8Pgc95we6EKl:-NI3ggJ80O3A_P55b6FW:b-evegkk) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc44f51e8993b3b49f8c218b4a43874a13a4c40f/pkg/server/tenant.go#L880)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Can you name me in this TODO instead? And it should reference NewGrantCoordinatorSQL instead -- which like you've seen, is not hooked up. Also, [nit] the `= nil` is unnecessary
</blockquote></details>

Done. This `nil` is necessary though. Otherwise, the `nil` check in `pacerFactory.NewPacer` fails. See [here](https://cockroachlabs.slack.com/archives/CLQ753JGH/p1668106423972389)


<!-- Sent from Reviewable.io -->

jayshrivastava commented:
bors r+

shermanCRL commented:
We should default this on for v23.1 (master) and default it off for v22.2 backport.
